sdk-tools/src/edger8r.rs

```diff
@@ -0,0 +1,368 @@
+// Copyright (c) 2018-2023 The MobileCoin Foundation
+
+#![doc = include_str!("../README.md")]
+#![deny(missing_docs, missing_debug_implementations, unsafe_code)]
+
+//! Edger8r Tool Wrapper
+
+use displaydoc::Display;
+use std::{
+    borrow::ToOwned,
+    env,
+    io::Error as IoError,
+    path::{Path, PathBuf},
+    process::Command,
+    string::String,
+};
+
+/// Errors which can occur when working with the edger8r tool.
+#[derive(Debug, Display)]
+pub enum Error {
+    /// There was an error running the command: {0}
+    Io(IoError),
+    /**
+     * There was an error generating the code,
+     * command:\n{0}\nstdout:\n{1}\n\nstderr:\n{2}
+     */
+    Generate(String, String, String),
+}
+
+impl From<IoError> for Error {
+    fn from(src: IoError) -> Error {
+        Error::Io(src)
+    }
+}
+
+/// The kind of C files to generate with the edger8r tool.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Default)]
+pub enum OutputKind {
+    /// Generate code for untrusted
+    Untrusted,
+    /// Generate code for trusted
+    Trusted,
+    /// Generate code for both trusted and untrusted
+    #[default]
+    Both,
+}
+
+/// Wrapper for the edger8r tool.
+///
+/// The Edger8r tool ships as part of the Intel® Software Guard Extensions SDK.
+/// It generates edge routines by reading a user-provided Enclave Description
+/// Language (EDL) file. These edge routines define the interface between the
+/// untrusted application and the enclave.
+///
+/// See [Edger8r Tool Documentation](https://download.01.org/intel-sgx/sgx-linux/2.18/docs/Intel_SGX_Developer_Reference_Linux_2.18_Open_Source.pdf#%5B%7B%22num%22%3A84%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C94.5%2C406.5%2C0%5D)
+/// for more details.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Edger8r {
+    /// The path to the `sgx_edger8r` executable.
+    edger8r_path: PathBuf,
+    /// The output directory to generate the code files into
+    out_dir: Option<PathBuf>,
+    /// The path to the EDL file to generate code for.
+    edl_file: PathBuf,
+    /// The type of code to be generated.
+    output_kind: OutputKind,
+}
+
+/// The files generated by the edger8r tool
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Edger8rFiles {
+    /// The trusted files generated. Will be empty when
+    /// [`OutputKind::Untrusted`] is used.
+    pub trusted: Vec<PathBuf>,
+    /// The untrusted files generated. Will be empty when
+    /// [`OutputKind::Trusted`] is used.
+    pub untrusted: Vec<PathBuf>,
+}
+
+impl Edger8r {
+    /// Create a new edger8r builder.
+    ///
+    /// # Arguments
+    /// * `edl_file` - The edl file to generate C files from
+    pub fn new(edl_file: impl AsRef<Path>) -> Self {
```

Comment:
Why is this not returning a result with and error field like in https://github.com/mobilecoinfoundation/mobilecoin/blob/master/util/build/sgx/src/edger8r.rs#L101 ?

Reply:
The main MobileCoin repo was using a result because of this line
```rust
let edger8r_path = get_binary(env.target_arch(), "sgx_edger8r")?;
```
This implementation utilizes [`sgx_bin_x64_dir()`](https://docs.rs/mc-sgx-core-build/0.4.0/mc_sgx_core_build/fn.sgx_bin_x64_dir.html) which doesn't fail. It returns either the `SGX_SDK` env variable or `opt/intel/sgxsdk` for the path root.
The `Command` will fail in `generate()` with "No such file or directory" if the sgx_edger8r path doesn't actually exist, https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d32be91dce2143444fa1d469ac5a4b9c